internal/shared/git/git.go

```diff
@@ -0,0 +1,200 @@
+
+// Parse parses rawurl into a URL structure. Parse first attempts to
+// find a standard URL with a valid Git transport as its scheme. If
+// that cannot be found, it then attempts to find a SCP-like URL. And
+// if that cannot be found, it assumes rawurl is a local path. If none
+// of these rules apply, Parse returns an error.
+func Parse(rawurl string) (u *url.URL, err error) {
+	parsers := []Parser{
+		ParseTransport,
+		ParseScp,
+		ParseLocal,
+	}
+
+	// Apply each parser in turn; if the parser succeeds, accept its
+	// result and return.
+	for _, p := range parsers {
+		u, err = p(rawurl)
+		if err == nil {
+			return u, err
+		}
+	}
+
+	// It's unlikely that none of the parsers will succeed, since
+	// ParseLocal is very forgiving.
+	return new(url.URL), fmt.Errorf("failed to parse %q", rawurl)
+}
+
+// ParseTransport parses rawurl into a URL object. Unless the URL's
+// scheme is a known Git transport, ParseTransport returns an error.
+func ParseTransport(rawurl string) (*url.URL, error) {
+	u, err := url.Parse(rawurl)
+	if err == nil && !Transports.Valid(u.Scheme) {
+		err = fmt.Errorf("scheme %q is not a valid transport", u.Scheme)
+	}
+	return u, err
+}
+
+// ParseScp parses rawurl into a URL object. The rawurl must be
+// an SCP-like URL, otherwise ParseScp returns an error.
+func ParseScp(rawurl string) (*url.URL, error) {
+	match := scpSyntax.FindAllStringSubmatch(rawurl, -1)
+	if len(match) == 0 {
+		return nil, fmt.Errorf("no scp URL found in %q", rawurl)
+	}
+	m := match[0]
+	user := strings.TrimRight(m[1], "@")
+	var userinfo *url.Userinfo
+	if user != "" {
+		userinfo = url.User(user)
+	}
+	rawquery := ""
+	if len(m) > 3 {
+		rawquery = m[4]
+	}
+	return &url.URL{
+		Scheme:   "ssh",
+		User:     userinfo,
+		Host:     m[2],
+		Path:     m[3],
+		RawQuery: rawquery,
+	}, nil
+}
+
+// ParseLocal parses rawurl into a URL object with a "file"
+// scheme. This will effectively never return an error.
+func ParseLocal(rawurl string) (*url.URL, error) {
+	return &url.URL{
+		Scheme: "file",
+		Host:   "",
+		Path:   rawurl,
+	}, nil
+}
+
+func GetAuthMethod(ctx context.Context, repoCred *controllerconfig.GitCredential, kubeClient kubernetes.Client, namespace string) (transport.AuthMethod, error) {
+	var authMethod transport.AuthMethod
+	switch {
+	case repoCred.HTTPCredential != nil:
+		cred := repoCred.HTTPCredential
+		username := cred.Username
+		secret, err := kubeClient.GetSecret(ctx, namespace, cred.Password.Key)
+		if err != nil {
+			return nil, err
+		}
+		password := string(secret.Data["password"])
+		authMethod = &gitHttp.BasicAuth{
+			Username: username,
+			Password: password,
+		}
+
+	case repoCred.SSHCredential != nil:
+		key := repoCred.SSHCredential.SSHKey.Key
+		secret, err := kubeClient.GetSecret(ctx, namespace, key)
+		if err != nil {
+			return nil, err
+		}
+		authMethod, err = ssh.NewPublicKeys("git", secret.Data["sshKey"], "")
+		if err != nil {
+			return nil, err
+		}
+
+	case repoCred.TLS != nil:
```

Comment:
The way this should work is that you take whatever protocol the RepositoryPath is (http or ssh) and then see if there's a corresponding HTTPCredential or SSHCredential depending on that protocol. Then there may or may not be a TLSCredential in addition.

Reply:
Ok ,Acknowledged .I am addressing that in another PR

Reply:
In that case we should close this PR. Because this is wrong for this PR.